0000-encryption-api/README.md

```diff
@@ -0,0 +1,135 @@
+To simplify the interfaces for Ursa users, *nonce* generation will be handled by Ursa and not a required concern for developers.
+
+Authenticated encryption methods are strongly preferred and any other methods should be considered hazardous materials–only used by those who know what they are doing.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+All proposed code will be implemented in **libursa**.
+**Libursa** will use existing traits and structures where possible.
+`KeyGenOption`, `PublicKey`, `PrivateKey`, `SecretKey`, and `MacKey`
+are already implemented.
+
+Rust structs and traits will look and work like so:
+```rust
+pub trait PublicKeyEncryptionScheme {
+    fn new() -> Self;
+    fn keypair(&self, options: Option<KeyGenOption>) -> Result<(PublicKey, PrivateKey), CryptoError>;
+    fn seal(&self, plaintext: &[u8], pk: &PublicKey) -> Result<Vec<u8>, CryptoError>;
+    fn open(&self, ciphertext: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, CryptoError>;
+    fn private_key_size() -> usize;
+    fn public_key_size() -> usize;
+}
+
+pub trait SecretKeyEncryptionScheme {
+    fn new() -> Self;
+    fn genkey(&self, options: Option<KeyGenOption>) -> Result<SecretKey, CryptoError>;
+    fn seal(&self, plaintext: &[u8], sk: &SecretKey) -> Result<Vec<u8>, CryptoError>;
+    fn open(&self, ciphertext: &[u8], sk: &SecretKey) -> Result<Vec<u8>, CryptoError>;
+    fn key_size() -> usize;
+    fn tag_size() -> usize;
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Encryption could be handled via other methods to avoid algorithm proliferation.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+Other libraries tie encryption methods to key objects instead of algorithm objects.
+The disadvantage of this model becomes apparent when multiple keys must be used for encryption
+and when safeguarding them.
+
+The key object model requires creating a new object then calling the encryption on the object.
+Our model considers keys as data blobs and only need one instance of an encryption scheme which allows any
+key to be passed without creating a new instance.
+
+Safeguarding should be easier where the key is a data blob that can be stored in secure locations.
+Key objects tend to encode more than just data when serialized.
+
+Other crypto libraries like libsodium and secp256k1 also follow this model with openssl beginning to adopt it as well.
+
+# Prior art
+[prior-art]: #prior-art
+
+[Rust Crypto](https://crates.io/crates/rust-crypto) implements a number of crypto primitives but is not for the faint of heart.
+It also does not allow for optionally compiling selected algorithms. It doesn't cover other algorithms that are under consideration to be added.
+
+[Libsodium](https://libsodium.gitbook.io/doc/) only utilizes Ed25519 curves.
+
+[Openssl](https://www.openssl.org) is a library that has many cryptographic algorithms but the interfaces are complicated and considered hazardous material.
```

Comment:
Maybe replace "hazardous material" with a bit about how it contains a huge amount of things for backwards compatibility that aren't need for modern blockchains?

Reply:
I was also meaning that to use those interfaces requires a great deal of understanding the underlying code too.

Reply:
Sure.  Let's just not offend anyone with our description of their algorithms ;).